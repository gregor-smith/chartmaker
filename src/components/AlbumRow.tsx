import React, { FC } from 'react'
import { css } from 'emotion'

import { Album as AlbumDetails } from '../reducer'
import { Album } from './Album'


type Props = {
    albums: (AlbumDetails | null)[]
    size: number
    from?: number
    to?: number
}


const style = css({
    display: 'flex'
})


export const AlbumRow: FC<Props> = ({ albums, from = 0, to = albums.length, size }) => {
    const row: JSX.Element[] = []
    for (let index = from; index < to; index++) {
        const album = albums[index] ?? null
        row.push(<Album key={index} details={album} size={size}/>)
    }
    return (
        <div className={style}>
            {row}
        </div>
    )
}
