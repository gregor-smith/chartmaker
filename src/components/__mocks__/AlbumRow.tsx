import type { FC } from 'react'

import type { AlbumRowProps } from '@/components/AlbumRow'


export const AlbumRow: FC<AlbumRowProps> = ({ albums, highlighted, size }) => {
    const json = JSON.stringify(albums)
    return (
        <div className='mock-album-row'>
            {`Albums: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
