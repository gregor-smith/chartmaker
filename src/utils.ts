import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
