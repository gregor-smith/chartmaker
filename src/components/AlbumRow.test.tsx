import { render } from 'react-dom'

import { AlbumRow } from './AlbumRow.js'

import { RenderContainer } from '../test-utils/utils.js'


const container = new RenderContainer()


test('renders styled container and children', () => {
    render(
        <AlbumRow>
            Test children
        </AlbumRow>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
