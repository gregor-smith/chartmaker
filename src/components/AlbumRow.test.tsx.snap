// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders styled container and children 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-0"
  >
    Test children
  </div>
</div>
`;
